src/modules/musig/session_impl.h

```diff
@@ -0,0 +1,634 @@
+    secp256k1_keyagg_cache_internal cache_i;
+    secp256k1_musig_session_internal session_i;
+    secp256k1_scalar mu, e, s;
+    secp256k1_gej pkj;
+    secp256k1_ge nonce_pt[2];
+    secp256k1_gej rj;
+    secp256k1_gej tmp;
+    secp256k1_ge pkp;
+    int i;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(partial_sig != NULL);
+    ARG_CHECK(pubnonce != NULL);
+    ARG_CHECK(pubkey != NULL);
+    ARG_CHECK(keyagg_cache != NULL);
+    ARG_CHECK(session != NULL);
+
+    if (!secp256k1_musig_session_load(ctx, &session_i, session)) {
+        return 0;
+    }
+
+    /* Compute "effective" nonce rj = aggnonce[0] + b*aggnonce[1] */
+    /* TODO: use multiexp */
+    for (i = 0; i < 2; i++) {
+        if (!secp256k1_musig_pubnonce_load(ctx, nonce_pt, pubnonce)) {
+            return 0;
+        }
+    }
+    secp256k1_gej_set_ge(&rj, &nonce_pt[1]);
+    secp256k1_ecmult(&rj, &rj, &session_i.noncecoef, NULL);
+    secp256k1_gej_add_ge_var(&rj, &rj, &nonce_pt[0], NULL);
+
+    if (!secp256k1_xonly_pubkey_load(ctx, &pkp, pubkey)) {
+        return 0;
+    }
+    if (!secp256k1_keyagg_cache_load(ctx, &cache_i, keyagg_cache)) {
+        return 0;
+    }
+    /* Multiplying the messagehash by the KeyAgg coefficient is equivalent
+     * to multiplying the signer's public key by the coefficient, except
+     * much easier to do. */
+    secp256k1_musig_keyaggcoef(&mu, &cache_i, &pkp.x);
+    secp256k1_scalar_mul(&e, &session_i.challenge, &mu);
+
+    /* If the MuSig-aggregate point has an odd Y coordinate, the signers will
+     * sign for the negation of their individual xonly public key such that the
+     * aggregate signature is valid for the MuSig aggregated xonly key. If the
+     * MuSig-aggregate point was tweaked then `e` is negated if the aggregate key
+     * has an odd Y coordinate XOR the internal key has an odd Y coordinate.*/
+    if (secp256k1_fe_is_odd(&cache_i.pk.y)
+            != cache_i.internal_key_parity) {
+        secp256k1_scalar_negate(&e, &e);
+    }
+
+    if (!secp256k1_musig_partial_sig_load(ctx, &s, partial_sig)) {
+        return 0;
+    }
+    /* Compute -s*G + e*pkj + rj */
+    secp256k1_scalar_negate(&s, &s);
+    secp256k1_gej_set_ge(&pkj, &pkp);
+    secp256k1_ecmult(&tmp, &pkj, &e, &s);
+    if (session_i.fin_nonce_parity) {
+        secp256k1_gej_neg(&rj, &rj);
+    }
+    secp256k1_gej_add_var(&tmp, &tmp, &rj, NULL);
+
+    return secp256k1_gej_is_infinity(&tmp);
+}
+
+int secp256k1_musig_partial_sig_agg(const secp256k1_context* ctx, unsigned char *sig64, const secp256k1_musig_session *session, const secp256k1_musig_partial_sig * const* partial_sigs, size_t n_sigs) {
+    size_t i;
+    secp256k1_musig_session_internal session_i;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(sig64 != NULL);
+    ARG_CHECK(session != NULL);
+    ARG_CHECK(partial_sigs != NULL);
+
```

Comment:
```suggestion
    ARG_CHECK(n_sigs > 0);
```

Reply:
That's not something that the rust type system can protect from and the output is well defined (as opposed to an "infinity" pubkey in key aggregation), so I'm leaning towards thinking that this is unnecessary.

Reply:
Ah no, I had found this earlier and it has been fixed: https://github.com/rust-bitcoin/rust-secp256k1/pull/304

Reply:
Hm, my point is that KeyAgg with 0 keys is _not_ wrong but we have to ARG_CHECK it because we can't give a result otherwise (since infinity can't be represented by a secp pubkey). Anyway, I added the ARG_CHECK because I don't want to waste any more of our time on this.